README.md
# glicko2-lite

An implementation of the Glicko-2 rating algorithm written from scratch, with
the goal being to provide less conveniences than [`glicko2js`][1] in favor of
faster execution time.

Basically `glicko2-lite` does not keep track of anything for you. All it does
is calculate new ratings. Once I reached the point where I was trying to
calculate ratings for tens of thousands of players across tens of thousands of
matches, `glicko2js` only got slower and slower as more matches were added. For
my case at least, I was already saving rating information to a database so I had
no need for the `Player` classes that `glicko2js` provides; all I really needed
was the actual calculation of new ratings itself.

Let's look at some benchmarks. First, the simplest case: create two players,
play a match between them, and calculate the new rating, deviation, and
volatility of both players:

``` bash
$ node bench/single.js

Starting...
  3 tests completed.

  glicko2-lite x 1,719,122 ops/sec ±0.60% (92 runs sampled)
  glicko2      x 1,406,286 ops/sec ±0.26% (94 runs sampled)
  glicko2.ts   x   755,440 ops/sec ±0.16% (100 runs sampled)
```

Now something larger: let's create 1,000 players, and have the first player play
a match against each of the other 999 players. After each individual match,
calculate the new ratings of the first player and their opponent:

``` bash
$ node bench/many.js

Starting...
  3 tests completed.

  glicko2-lite x 1,457 ops/sec ±0.39% (94 runs sampled)
  glicko2      x 33.92 ops/sec ±1.45% (59 runs sampled)
  glicko2.ts   x  0.12 ops/sec ±1.52% (5 runs sampled)
```

Some might say this is unfair, since new ratings are being calculated piecemeal
after each match instead of in bulk after a series, or rating period, of
matches. Let's try the same benchmark but only calculating the new ratings at
the end of all of the matches:

``` bash
$ node bench/batch.js

Starting...
  3 tests completed.

  glicko2-lite x 21,306 ops/sec ±1.20% (92 runs sampled)
  glicko2      x  2,304 ops/sec ±0.78% (91 runs sampled)
  glicko2.ts   x    105 ops/sec ±0.72% (77 runs sampled)
```


   [1]: https://github.com/mmai/glicko2js

## Example

``` javascript
const glicko2 = require('glicko2-lite');

// player A: 1500 rating, 350 rating deviation, and 0.06 volatility
// player B: 2000 rating, 70 rating deviation
// A loses to B
// new rating, rating deviation, and volatility:
glicko2(1500, 350, 0.06, [[2000, 70, 0]])
// => {
// =>   rating: 1467.5878493169462,
// =>   rd: 318.6617548537152,
// =>   vol: 0.059999457650202655
// => }
```

## Installation

``` bash
$ npm install glicko2-lite
```

## API

``` javascript
const glicko2 = require('glicko2-lite');
```

### `glicko2(rating, rd, vol, matches, [options])`

  - `rating` (_Number_)
  - `rd` (_Number_)
  - `vol` (_Number_)
  - `matches` (_Array_)
  - `options` (_Object_)


CHANGELOG.md
## [4.0.0](https://github.com/kenany/glicko2-lite/compare/3.2.0...4.0.0) (2023-09-06)


### ⚠ BREAKING CHANGES

* Node.js v14 and v16 are no longer supported.

### Features

* drop Node.js v14 and v16 ([4b423ea](https://github.com/kenany/glicko2-lite/commit/4b423eaabf1a7ed1ee878059e64145dd4dba3c0f))

## [3.2.0](https://github.com/KenanY/glicko2-lite/compare/3.1.0...3.2.0) (2023-04-26)


### Features

* update inequality in step 5.4 ([72d4996](https://github.com/KenanY/glicko2-lite/commit/72d499643f6ec74b335ea2edd145392c1a478759))

## [3.1.0](https://github.com/KenanY/glicko2-lite/compare/3.0.0...3.1.0) (2023-03-26)


### Features

* add TypeScript declarations ([ef22831](https://github.com/KenanY/glicko2-lite/commit/ef22831edb591d0fce9a9c0bfc67cd2bd875afbf))
* generate TypeScript declarations from source ([c98b168](https://github.com/KenanY/glicko2-lite/commit/c98b1689065fd7d53414be9fbacc51c6ca71d6d3))

## [3.0.0](https://github.com/KenanY/glicko2-lite/compare/2.0.1...3.0.0) (2022-11-14)


### ⚠ BREAKING CHANGES

* Node.js v12 is no longer supported.

### Features

* drop Node.js v12 support ([f2c2395](https://github.com/KenanY/glicko2-lite/commit/f2c23958288569e1712fd23494412eef3083afc8))

### [2.0.1](https://github.com/KenanY/glicko2-lite/compare/2.0.0...2.0.1) (2021-11-24)


### Bug Fixes

* add type coverage through JSDoc ([1da099b](https://github.com/KenanY/glicko2-lite/commit/1da099b1f530f4f1f8e4bc77afaf154296d45ee5))

## [2.0.0](https://github.com/KenanY/glicko2-lite/compare/1.0.2...2.0.0) (2021-11-23)


### ⚠ BREAKING CHANGES

* Node.js v10 is no longer supported.

### Features

* drop Node.js v10 support ([2099792](https://github.com/KenanY/glicko2-lite/commit/209979202179756186bd111c0d9ac3b7917a8443))

### [1.0.2](https://github.com/KenanY/glicko2-lite/compare/1.0.1...1.0.2) (2021-11-23)


### Bug Fixes

* **deps:** lodash.assign@^4.2.0 ([13d5316](https://github.com/KenanY/glicko2-lite/commit/13d531607d872cb38de3dd3587c08f9ea3f367e3))
* **deps:** lodash.map@^4.6.0 ([f1f50f8](https://github.com/KenanY/glicko2-lite/commit/f1f50f8598bd6edff789939f02140e7c999c1cbd))
* **deps:** lodash.reduce@4.6.0 ([34de30a](https://github.com/KenanY/glicko2-lite/commit/34de30a5fb43372920f03e293e72597570e76e91))


test/index.js
'use strict';

const test = require('tape');
const isFunction = require('lodash.isfunction');
const almostEqual = require('almost-equal');

const glicko2 = require('../');

test('exports a function', (t) => {
  t.plan(1);
  t.ok(isFunction(glicko2));
});

test('calculates new ratings', (t) => {
  t.plan(3);

  let a = { rating: 1500, rd: 200, vol: 0.06 };
  const b = { rating: 1400, rd: 30, vol: 0.06 };
  const c = { rating: 1550, rd: 100, vol: 0.06 };
  const d = { rating: 1700, rd: 300, vol: 0.06 };

  // a beats b
  // c beats a
  // d beats a
  a = glicko2(a.rating, a.rd, a.vol, [
    [b.rating, b.rd, 1],
    [c.rating, c.rd, 0],
    [d.rating, d.rd, 0]
  ], { tau: 0.5 });

  t.ok(almostEqual(a.rating, 1464, 0.1), a.rating + ' ~= 1464');
  t.ok(almostEqual(a.rd, 151.52, 0.01), a.rd + ' ~= 151.52');
  t.ok(almostEqual(a.vol, 0.05999, 0.00001), a.vol + ' ~= 0.05999');
});


index.js
'use strict';

/** @typedef {[rating: number, rd: number, vol: number]} Opponent */

/**
 * @typedef {object} ScaledOpponent
 * @property {number} muj
 * @property {number} phij
 * @property {number} gphij
 * @property {number} emmp
 * @property {number} score
 */

/**
 * @param {number} rating
 * @param {number} rd
 * @param {{ rating: number; }} options
 */
function scale(rating, rd, options) {
  const mu = (rating - options.rating) / 173.7178;
  const phi = rd / 173.7178;
  return { mu, phi };
}

/**
 * @param {number} phi
 * @returns {number}
 */
function g(phi) {
  return 1 / Math.sqrt(1 + 3 * Math.pow(phi, 2) / Math.pow(Math.PI, 2));
}

/**
 * @param {number} mu
 * @param {number} muj
 * @param {number} phij
 * @returns {number}
 */
function e(mu, muj, phij) {
  return 1 / (1 + Math.exp(-g(phij) * (mu - muj)));
}

/**
 * @param {number} mu
 * @param {readonly Opponent[]} opponents
 * @param {{ rating: number; }} options
 * @returns {ScaledOpponent[]}
 */
function scaleOpponents(mu, opponents, options) {
  return opponents.map((opp) => {
    const scaled = scale(opp[0], opp[1], options);
    return {
      muj: scaled.mu,
      phij: scaled.phi,
      gphij: g(scaled.phi),
      emmp: e(mu, scaled.mu, scaled.phi),
      score: opp[2]
    };
  });
}

/**
 * @param {readonly ScaledOpponent[]} opponents
 * @returns {number}
 */
function updateRating(opponents) {
  return 1 / opponents.reduce(
    (sum, opp) => sum + Math.pow(opp.gphij, 2) * opp.emmp * (1 - opp.emmp),
    0
  );
}

/**
 * @param {number} v
 * @param {readonly ScaledOpponent[]} opponents
 * @returns {number}
 */
function computeDelta(v, opponents) {
  return v * opponents.reduce(
    (sum, opp) => sum + opp.gphij * (opp.score - opp.emmp),
    0
  );
}

/**
 * @param {number} phi
 * @param {number} v
 * @param {number} delta
 * @param {number} a
 * @param {{ tau: number; }} options
 */
function volF(phi, v, delta, a, options) {
  const phi2 = Math.pow(phi, 2);
  const d2 = Math.pow(delta, 2);

  /**
   * @param {number} x
   */
  return function(x) {
    const ex = Math.exp(x);
    const a2 = phi2 + v + ex;
    const p2 = (x - a) / Math.pow(options.tau, 2);
    const p1 = (ex * (d2 - phi2 - v - ex)) / (2 * Math.pow(a2, 2));
    return p1 - p2;
  };
}

/**
 * @param {number} sigma
 * @param {number} phi
 * @param {number} v
 * @param {number} delta
 * @param {{ tau: number; }} options
 * @returns {number}
 */
function computeVolatility(sigma, phi, v, delta, options) {
  // 5.1
  let a = Math.log(Math.pow(sigma, 2));
  const f = volF(phi, v, delta, a, options);

  // 5.2
  /** @type {number} */
  let b;
  if (Math.pow(delta, 2) > Math.pow(phi, 2) + v) {
    b = Math.log(Math.pow(delta, 2) - Math.pow(phi, 2) - v);
  }
  else {
    let k = 1;
    while (f(a - k * options.tau) < 0) {
      k++;
    }
    b = a - k * options.tau;
  }

  // 5.3
  let fa = f(a);
  let fb = f(b);

  // 5.4
  while (Math.abs(b - a) > 0.000001) {
    /** @type {number} */
    const c = a + (a - b) * fa / (fb - fa);
    const fc = f(c);

    if (fc * fb <= 0) {
      a = b;
      fa = fb;
    }
    else {
      fa /= 2;
    }

    b = c;
    fb = fc;
  }

  // 5.5
  return Math.exp(a / 2);
}

/**
 * @param {number} sigmap
 * @param {number} phi
 * @returns {number}
 */
function phiStar(sigmap, phi) {
  return Math.sqrt(Math.pow(sigmap, 2) + Math.pow(phi, 2));
}

/**
 * @param {number} phis
 * @param {number} mu
 * @param {number} v
 * @param {readonly ScaledOpponent[]} opponents
 */
function newRating(phis, mu, v, opponents) {
  const phip = 1 / Math.sqrt(1 / Math.pow(phis, 2) + 1 / v);
  const mup = mu + Math.pow(phip, 2) * opponents.reduce(
    (sum, opp) => sum + opp.gphij * (opp.score - opp.emmp),
    0
  );
  return { mu: mup, phi: phip };
}

/**
 * @param {number} mup
 * @param {number} phip
 * @param {{ rating: number; }} options
 */
function unscale(mup, phip, options) {
  const rating = 173.7178 * mup + options.rating;
  const rd = 173.7178 * phip;
  return { rating, rd };
}

/**
 * @param {number} rating Rating.
 * @param {number} rd Rating deviation.
 * @param {number} sigma Rating volatility.
 * @param {readonly Opponent[]} opponents Opponents.
 * @param {{ rating?: number; tau?: number; }} [options] Options.
 * @returns {{ rating: number; rd: number; vol: number; }}
 */
function rate(rating, rd, sigma, opponents, options) {
  /** @type {{ rating: number; tau: number; }} */
  const opts = Object.assign({}, { rating: 1500, tau: 0.5 }, options || {});

  // Step 2
  const scaled = scale(rating, rd, opts);
  const scaledOpponents = scaleOpponents(scaled.mu, opponents, opts);

  // Step 3
  const v = updateRating(scaledOpponents);

  // Step 4
  const delta = computeDelta(v, scaledOpponents);

  // Step 5
  const sigmap = computeVolatility(sigma, scaled.phi, v, delta, opts);

  // Step 6
  const phis = phiStar(sigmap, scaled.phi);

  // Step 7
  const updated = newRating(phis, scaled.mu, v, scaledOpponents);

  return Object.assign(
    {},
    unscale(updated.mu, updated.phi, opts),
    { vol: sigmap }
  );
}

module.exports = rate;


bench/many.js
'use strict';

const Benchmark = require('benchmark');
const benchmarks = require('beautify-benchmark');
const glicko2 = require('glicko2');
const glicko2ts = require('glicko2.ts');

const lite = require('../');

const suite = new Benchmark.Suite();

// Benchmark:
//   1. Create 1000 players.
//   2. Have player 1 play a match versus every other player, one at a time.
//   3. Get the new rating, deviation, and volatility of player 1.

/* eslint-disable no-console */

suite.add('glicko2-lite', () => {
  const players = [];

  let i = 0;
  while (i++ < 1000) {
    players.push({ rating: 1500, rd: 350, vol: 0.06 });
  }

  for (let i = 1; i < players.length; i++) {
    const p1 = players[0];
    const p2 = players[i];
    players[0] = lite(
      p1.rating,
      p1.rd,
      p1.vol,
      [[p2.rating, p2.rd, Math.floor(Math.random() * 3) / 2]],
      { tau: 0.5 }
    );
    players[i] = lite(
      p2.rating,
      p2.rd,
      p2.vol,
      [[p1.rating, p1.rd, Math.floor(Math.random() * 3) / 2]],
      { tau: 0.5 }
    );
  }

  return players[0];
});

suite.add('glicko2', () => {
  const glicko = new glicko2.Glicko2({ tau: 0.5 });

  const players = [];

  let i = 0;
  while (i++ < 1000) {
    players.push(glicko.makePlayer(1500, 350, 0.06));
  }

  const p1 = players[0];

  for (let i = 1; i < players.length; i++) {
    const p2 = players[i];
    glicko.updateRatings([[p1, p2, Math.floor(Math.random() * 3) / 2]]);
  }

  return { rating: p1.getRating(), rd: p1.getRd(), vol: p1.getVol() };
});

suite.add('glicko2.ts', () => {
  const glicko = new glicko2ts.Glicko2({ tau: 0.5 });

  const players = [];

  let i = 0;
  while (i++ < 1000) {
    players.push(glicko.makePlayer(1500, 350, 0.06));
  }

  const p1 = players[0];

  for (let i = 1; i < players.length; i++) {
    const p2 = players[i];
    glicko.updateRatings([[p1, p2, Math.floor(Math.random() * 3) / 2]]);
  }

  return { rating: p1.getRating(), rd: p1.getRd(), vol: p1.getVol() };
});

suite.on('cycle', (event) => {
  benchmarks.add(event.target);
});

suite.on('start', () => {
  console.log('Starting...');
});

suite.on('complete', () => {
  benchmarks.log();
});

suite.run({ async: false });


bench/batch.js
// @ts-check

'use strict';

const Benchmark = require('benchmark');
const benchmarks = require('beautify-benchmark');
const glicko2 = require('glicko2');
const glicko2ts = require('glicko2.ts');

const lite = require('../');

const suite = new Benchmark.Suite();

// Benchmark:
//   1. Create 1000 players.
//   2. Have player 1 play a match versus every other player, in a single batch.
//   3. Get the new rating, deviation, and volatility of player 1.

/* eslint-disable no-console */

suite.add('glicko2-lite', () => {
  const players = [];

  /** @type {lite.Opponent[]} */
  const matches = [];

  let i = 0;
  while (i++ < 1000) {
    const player = { rating: 1500, rd: 350, vol: 0.06 };
    players.push(player);
    matches.push([player.rating, player.rd, Math.floor(Math.random() * 3) / 2]);
  }

  return lite(
    players[0].rating,
    players[0].rd,
    players[0].vol,
    matches,
    { tau: 0.5 }
  );
});

suite.add('glicko2', () => {
  const glicko = new glicko2.Glicko2({ tau: 0.5 });

  const p1 = glicko.makePlayer(1500, 350, 0.06);
  const players = [p1];
  const matches = [];

  let i = 1;
  while (i++ < 1000) {
    const player = glicko.makePlayer(1500, 350, 0.06);
    players.push(player);
    matches.push([p1, player, Math.floor(Math.random() * 3) / 2]);
  }

  glicko.updateRatings(matches);

  return { rating: p1.getRating(), rd: p1.getRd(), vol: p1.getVol() };
});

suite.add('glicko2.ts', () => {
  const glicko = new glicko2ts.Glicko2({ tau: 0.5 });

  const p1 = glicko.makePlayer(1500, 350, 0.06);
  const players = [p1];

  /** @type {glicko2ts.playerMatch[]} */
  const matches = [];

  let i = 1;
  while (i++ < 1000) {
    const player = glicko.makePlayer(1500, 350, 0.06);
    players.push(player);
    matches.push([p1, player, Math.floor(Math.random() * 3) / 2]);
  }

  glicko.updateRatings(matches);

  return { rating: p1.getRating(), rd: p1.getRd(), vol: p1.getVol() };
});

suite.on('cycle', (event) => {
  benchmarks.add(event.target);
});

suite.on('start', () => {
  console.log('Starting...');
});

suite.on('complete', () => {
  benchmarks.log();
});

suite.run({ async: false });


bench/single.js
'use strict';

const Benchmark = require('benchmark');
const benchmarks = require('beautify-benchmark');
const glicko2 = require('glicko2');
const glicko2ts = require('glicko2.ts');

const lite = require('../');

const suite = new Benchmark.Suite();

// Benchmark:
//   1. Create two players.
//   2. Have player 1 win a match against player 2.
//   3. Get the new rating, deviation, and volatility of both players.

/* eslint-disable no-console, no-unused-vars */

suite.add('glicko2-lite', () => {
  // 1. Create two players.
  const player1 = { rating: 1500, rd: 350, vol: 0.06 };
  const player2 = { rating: 1500, rd: 350, vol: 0.06 };

  // 2. Have player 1 win a match against player 2.
  // 3. Get the new rating, deviation, and volatility of both players.
  const new1 = lite(
    player1.rating,
    player1.rd,
    player1.vol,
    [[player2.rating, player2.rd, 1]],
    { tau: 0.5 }
  );
  const new2 = lite(
    player2.rating,
    player2.rd,
    player2.vol,
    [[player1.rating, player1.rd, 0]],
    { tau: 0.5 }
  );
});

suite.add('glicko2', () => {
  // 1. Create two players.
  const glicko = new glicko2.Glicko2({ tau: 0.5 });
  const player1 = glicko.makePlayer(1500, 350, 0.06);
  const player2 = glicko.makePlayer(1500, 350, 0.06);

  // 2. Have player 1 win a match against player 2.
  glicko.updateRatings([[player1, player2, 1]]);

  // 3. Get the new rating, deviation, and volatility of both players.
  const new1 = {
    rating: player1.getRating(),
    rd: player1.getRd(),
    vol: player1.getVol()
  };
  const new2 = {
    rating: player2.getRating(),
    rd: player2.getRd(),
    vol: player2.getVol()
  };
});

suite.add('glicko2.ts', () => {
  // 1. Create two players.
  const glicko = new glicko2ts.Glicko2({ tau: 0.5 });
  const player1 = glicko.makePlayer(1500, 350, 0.06);
  const player2 = glicko.makePlayer(1500, 350, 0.06);

  // 2. Have player 1 win a match against player 2.
  glicko.updateRatings([[player1, player2, 1]]);

  // 3. Get the new rating, deviation, and volatility of both players.
  const new1 = {
    rating: player1.getRating(),
    rd: player1.getRd(),
    vol: player1.getVol()
  };
  const new2 = {
    rating: player2.getRating(),
    rd: player2.getRd(),
    vol: player2.getVol()
  };
});

suite.on('cycle', (event) => {
  benchmarks.add(event.target);
});

suite.on('start', () => {
  console.log('Starting...');
});

suite.on('complete', () => {
  benchmarks.log();
});

suite.run({ async: false });
